s2i-liberty

raffaelespazzoli opened the pull request.

etsauer commented on s2i-liberty/Dockerfile:
```diff
@@ -0,0 +1,38 @@
+# This image provides a base for building and running WildFly applications.
+# It builds using maven and runs the resulting artifacts on WildFly 10.1.0 Final
+
+FROM websphere-liberty:webProfile7
```
Is this using a red hat based base or debian?

raffaelespazzoli replied:
the builder image can be any. The runtime image is created by IBM based on debian. I changed the user to our standard 1001. I think we want to keep the IBM's provided image as our base image because it increases the chances for our customers that IBM will support it.

etsauer commented on s2i-liberty/readme.md:
````diff
@@ -0,0 +1,197 @@
+The content used to produce the Liberty runtime image can originate from a Git repository. Execute the following command to start a new image build using the git source strategy.:
+
+```
+oc new-build websphere-liberty:webProfile7~https://github.com/raffaelespazzoli/containers-quickstarts --context-dir=s2i-liberty --name=liberty-runtime-s2i --strategy=docker
````
let's change the git repo name to `https://github.com/redhat-cop/containers-quickstarts`

etsauer commented on s2i-liberty/readme.md:
```diff
@@ -0,0 +1,197 @@
+
+In this example we will use a pre-existing builder image that can build maven based apps: `registry.access.redhat.com/jboss-eap-7/eap70-openshift` 
+
+#### Git Source
```
So this heading to me implies "Hey let's build an app from source using our liberty base image", which in actuality, this is the step to `docker build` the base image in OpenShift. I think we should change this heading to "Building the S2I Base Image", and then have a separate "Build from Git" heading that shows the `oc new-app` command pointing to an application git repo, in addition to the "Binary Build" steps below.

raffaelespazzoli replied:
I didn't get this comment. can we talk about it? I applied the other feedback

etsauer replied:
We can disregard this comment

sabre1041 commented on s2i-liberty/readme.md:
```diff
@@ -0,0 +1,197 @@
+
+OpenShift provides several out of the box Source to Image builder images. To support deployments in Liberty, a new s2i builder image will be created to support a simplified deployment to OpenShift. Once the new image is produced, an example application will be deployed. 
+
+We will use the [extended builds approach](https://docs.openshift.org/latest/dev_guide/builds.html#extended-builds).
```
This should reference the OCP docs instead of origin

https://docs.openshift.com/container-platform/3.3/dev_guide/builds.html#extended-builds

sabre1041 commented on s2i-liberty/readme.md:
````diff
@@ -0,0 +1,197 @@
+        - sourcePath: /opt/eap/standalone/deployments/hello-world-war-1.0.0.war
+          destinationDir: artifacts
+```
+notice that the source strategy has two additional sections:
````
Can we provide this modifications as a _patch_ command that users can easily apply?

sabre1041 commented on s2i-liberty/readme.md:
```diff
@@ -0,0 +1,197 @@
+
+### Create a new Application
+
+To demonstrate the usage of the newly created builder and runtime images, a Jee example application will be built and deployed to Liberty using the Source to Image process. 
```
Capitalize JEE

sabre1041 commented on s2i-liberty/readme.md:
````diff
@@ -0,0 +1,197 @@
+The build that was triggered by the `new-app` command can be found by executing the following command:
+
+```
+oc get builds -l=app=play-app
````
Change to hello-world assuming that is the correct label

sabre1041 commented on s2i-liberty/readme.md:
```diff
@@ -0,0 +1,197 @@
+
+* `oc new-app` - OpenShift command to create a a new application
+* `-i=hello-world` - Name of the ImageStream that contains the result of the build config that uses the extended s2i process
+* `--name=play-app` - Name to be applied to the newly created resources
```
Change to hello-world

sabre1041 commented on s2i-liberty/readme.md:
```diff
@@ -0,0 +1,197 @@
+
+## Considerations on http session failover
+
+When liberty is deployed in a cloud environment certain limitation applies as explained in this [document](http://www.ibm.com/support/knowledgecenter/en/SSD28V_8.5.5/com.ibm.websphere.wlp.core.doc/ae/cwlp_paas_restrict.html).
```
Sentence needs to be cleaned up. "When liberty is deployed in a cloud environment, there are certain limitations that apply as explained in this document"

etsauer commented:
@raffaelespazzoli this looks much better. However, after I run the `oc new-app` to create the deployment, the pod gets stuck in a CrashLoop with the following error:
```
[esauer@localhost containers-quickstarts]$ oc get pods
NAME                            READY     STATUS             RESTARTS   AGE
hello-world-1-build             0/1       Completed          0          14h
hello-world-1-mih0g             0/1       CrashLoopBackOff   2          14h
hello-world-artifacts-1-build   0/1       Completed          0          14h
s2i-liberty-1-build             0/1       Completed          0          14h
[esauer@localhost containers-quickstarts]$ oc logs -f hello-world-1-mih0g
JVMSHRC155E Error copying username into cache name
JVMSHRC686I Failed to startup shared class cache. Continue without using it as -Xshareclasses:nonfatal is specified
CWWKE0005E: The runtime environment could not be launched.
CWWKE0044E: There is no write permission for server directory /opt/ibm/wlp/output/defaultServer
```

raffaelespazzoli commented:
@etsauer I tried to reproduce the error, but it works fine with me. Which approach are you trying? Also can you make sure you start from a clean project?

sabre1041 commented:
@raffaelespazzoli The issue is most likely caused by the relaxed permissions of the Red Hat CDK. By default, all containers are run with the `anyuid` scc. When I ran through the steps in the PR using the baseline CDK configuration, I was able to launch the container successfully (as it was running with the `anyuid` scc). When I removed that privilege, with the following command, I hit the same error as @etsauer 

```
oc adm policy remove-scc-from-group anyuid system:authenticated
```

etsauer commented:
@sabre1041 what would cause that? the dockerfile uses `USER 1001` which should get a proper uuid right?

sabre1041 commented:
@etsauer no. by default in non-CDK installations, a user is randomly assigned from a range and the UID in the dockerfile is ignored as governed by the `restricted` scc. In the CDK, all authenticated users are given access to the `anyuid` scc.

raffaelespazzoli commented:
I added some instructions to the docker file which should fix the issue.
The image still works for me but I don't have an easily accessible OCP (I'm working on it).
@etsauer do you mind retesting it?

sabre1041 commented:
@raffaelespazzoli validated latest updates allow container to start up successfully.

README needs to add instructions on how to view the sample application. This requires the service to be exposed and in particular, the correct port (default is 7777, but application is available on 9080). In addition, application is available at `/hello-world-war-1.0.0` instead of `/`

raffaelespazzoli commented:
@sabre1041 thanks, I addressed the port issue and added instructions on how to verify the application.
I didn't change the context root of the application as that would require a change in the application code which I don't own.

sabre1041 commented:
@raffaelespazzoli this looks great. Would you be able to squash all of the commits? Then we should be good to merge

sabre1041 commented:
@etsauer you will need to approve changes before we will eventually be able to merge
